### PR TITLE
T/209: Upgraded dependencies

### LINKED ISSUE #209: Update dependencies of packages
They are quite outdated:

image

### LINKED ISSUE #236: Update sinon to the latest version
Sinon is going to change its API and update to `v2.3.5` produces a huge amount o warnings about uses of deprecated functions.

Most popular warning is about deprecated 3-arguments stubs `sinon.stub(x,y,z)`.
```
INFO: 'sandbox.stub(obj, 'meth', val) is deprecated and will be removed from the public API in a future version of sinon.
 Use sandbox.stub(obj, 'meth').callsFake(fn) instead in order to stub a function.
 Use sandbox.stub(obj, 'meth').value(fn) instead in order to stub a non-function value.'
INFO: 'sinon.stub(obj, 'meth', fn) is deprecated and will be removed from the public API in a future version of sinon.
 Use stub(obj, 'meth').callsFake(fn).
 Codemod available at https://github.com/hurrymaplelad/sinon-codemod'
```

### LINKED ISSUE #237: Update chai
Chai has changed an API and using latest `4.0.2` throws few errors in CKE5 tests.

I think most of errors is about dropped shorthand assertions like.
```
expect().to.object -> expect().to.be.an( 'object' )
```

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Internal: Upgraded dependencies. Closes #209. Closes #236. Closes #237.

### CODE REVIEW

**Comment by pomek** [comment]
R- because of some issues with CI occurred.

**Comment by pomek** [comment]
Ok, ready.

**Comment by Reinmar** [comment]
You missed a couple of packages:

image

**Comment by Reinmar** [comment]
I'll try to update these myself.

**Comment by Reinmar** [comment]
OK, I'm predicting a problem. Chain and Sinon are installed by ckeditor5-dev-tests. The problem is that they are used both for that package's tests and for the CKE5 tests. So when we'll update them they will crash CKE5's tests. Am I right?

**Comment by pomek** [comment]
I hope no because Sinon and Chain are defined in devDependencies. But I'm not sure…

**Comment by Reinmar** [comment]
Yes, I was right. After I updated these dependencies so ckeditor5-dev-tests uses the same version of Chai and Sinon in prod and dev modes the CKE5 tests start to throw.

Currently, both packages are listed twice in `package.json` which is bad because I'm often using dev versions of ckeditor5-dev-* packages in ckeditor5. We need these versions to be consistent.

So:

1. I pushed a commit here updating the remaining dependencies and deduping them.
2. You need to fix ckeditor5-dev-* tests because some of them log Sinon warnings.
3. You need to actually test your changes in ckeditor5. I think that the easiest way to do that in a reliable way is to:
   ```
   cd ckeditor5
   # add "../ckeditor5-dev/packages" to lerna.json so Lerna symlinks packages installed in 
   ckeditor5 to those available in this repo.
   lerna bootstrap
   ```
4. Test all gulp tasks which you can test (so all except the release task).

Only then this PR will make sense.

**Comment by pomek** [comment]
> You need to fix ckeditor5-dev-* tests because some of them log Sinon warnings.

Done in eb50099.

> Test all gulp tasks which you can test (so all except the release task).

* [x] lint
* [x] lint-staged
* [x] pre-commit
* [x] test
* [x] test:manual
* [x] docs
* [x] docs:api
* [ ] translations:collect
* [ ] translations:upload
* [ ] translations:download
* [x] changelog:dependencies
* [x] release:dependencies

Because I created the release tool tasks, I know how and when I can break the process without any effects. But with the translations, I don't even want to call these tasks.

After calling `npm t` in the main repository, 10 tests fail and a lot of them log Sinon warnings. These warnings are easy to fix but the fail tests can be hard to fix.

File: packages/ckeditor5-dev-utils/tests/workspace.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ describe( 'utils', () => {
 			it( 'should return only ckeditor5 directories', () => {
 				const workspacePath = '/workspace/path';
 				const sourceDirectories = [ 'tools', 'ckeditor5', 'ckeditor5-core', '.bin', 'ckeditor5-plugin-image' ];
-				sandbox.stub( tools, 'getDirectories', () => sourceDirectories );
+				sandbox.stub( tools, 'getDirectories' ).callsFake( () => sourceDirectories );
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-utils/tests/workspace.js]: It can be just `.returns( sourceDirectories )`. Am I right?

**Comment by pomek** [comment]
I fixed all the tests. After merging this PR, would be nice to read the migration guide [**from 1.x to 2.x**](http://sinonjs.org/releases/v3.0.0/migrating-to-2.0/) and from [**2.x to 3.x**](http://sinonjs.org/releases/v3.0.0/migrating-to-3.0/).

Along with the changes in this PR, the following branches in the repositories have to merged:
* [`ckeditor5-adapter-ckfinder`](https://github.com/ckeditor/ckeditor5-adapter-ckfinder/compare/t/ckeditor5-dev/209),
* [`ckeditor5-core`](https://github.com/ckeditor/ckeditor5-core/compare/t/ckeditor5-dev/209),
* [`ckeditor5-editor-balloon-toolbar`](https://github.com/ckeditor/ckeditor5-editor-balloon-toolbar/compare/t/ckeditor5-dev/209),
* [`ckeditor5-engine`](https://github.com/ckeditor/ckeditor5-engine/compare/t/ckeditor5-dev/209),
* [`ckeditor5-link`](https://github.com/ckeditor/ckeditor5-link/compare/t/ckeditor5-dev/209),
* [`ckeditor5-ui`](https://github.com/ckeditor/ckeditor5-ui/compare/t/ckeditor5-dev/209),
* [`ckeditor5-upload`](https://github.com/ckeditor/ckeditor5-upload/compare/t/ckeditor5-dev/209),
* [`ckeditor5-utils`](https:/github.com/ckeditor/ckeditor5-utils/compare/t/ckeditor5-dev/209),
* [`ckeditor5-widget`](https://github.com/ckeditor/ckeditor5-widget/compare/t/ckeditor5-dev/209).

Unfortunately, upgrade Sinon to the latest version caused one error in our test. We cannot stub `Array.prototype.filter` ([see](https://github.com/ckeditor/ckeditor5-utils/compare/t/ckeditor5-dev/209#diff-a70fd7d1713c585aa86f3921ab942875R480)) because of `Maximum call stack size exceeded`. I reported this bug – https://github.com/sinonjs/sinon/issues/1521.

**Comment by Reinmar** [comment]
> Unfortunately, upgrade Sinon to the latest version caused one error in our test. We cannot stub Array.prototype.filter (see) because of Maximum call stack size exceeded. I reported this bug – sinonjs/sinon#1521.

So does this test pass now or not? If not, then we need to look for workaround. We cannot leave it like this.

**Comment by pomek** [comment]
> So does this test pass now or not? If not, then we need to look for workaround. We cannot leave it like this.

All works fine. I did a workaround.

**Comment by pomek** [comment]
image

**Comment by Reinmar** [comment]
To be reviewed:

 * [ ] [`ckeditor5-adapter-ckfinder`](https://github.com/ckeditor/ckeditor5-adapter-ckfinder/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-core`](https://github.com/ckeditor/ckeditor5-core/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-editor-balloon-toolbar`](https://github.com/ckeditor/ckeditor5-editor-balloon-toolbar/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-engine`](https://github.com/ckeditor/ckeditor5-engine/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-link`](https://github.com/ckeditor/ckeditor5-link/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-ui`](https://github.com/ckeditor/ckeditor5-ui/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-upload`](https://github.com/ckeditor/ckeditor5-upload/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-utils`](https://github.com/ckeditor/ckeditor5-utils/compare/t/ckeditor5-dev/209),
 * [ ] [`ckeditor5-widget`](https://github.com/ckeditor/ckeditor5-widget/compare/t/ckeditor5-dev/209).

**Comment by pomek** [comment]
`mgit.json`

```json
{
  "packages": "packages/",
  "dependencies": {
    "@ckeditor/ckeditor5-adapter-ckfinder": "ckeditor/ckeditor5-adapter-ckfinder#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-autoformat": "ckeditor/ckeditor5-autoformat",
    "@ckeditor/ckeditor5-basic-styles": "ckeditor/ckeditor5-basic-styles",
    "@ckeditor/ckeditor5-block-quote": "ckeditor/ckeditor5-block-quote",
    "@ckeditor/ckeditor5-build-classic": "ckeditor/ckeditor5-build-classic",
    "@ckeditor/ckeditor5-clipboard": "ckeditor/ckeditor5-clipboard",
    "@ckeditor/ckeditor5-core": "ckeditor/ckeditor5-core#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-editor-balloon-toolbar": "ckeditor/ckeditor5-editor-balloon-toolbar#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-editor-classic": "ckeditor/ckeditor5-editor-classic",
    "@ckeditor/ckeditor5-editor-inline": "ckeditor/ckeditor5-editor-inline",
    "@ckeditor/ckeditor5-engine": "ckeditor/ckeditor5-engine#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-enter": "ckeditor/ckeditor5-enter",
    "@ckeditor/ckeditor5-heading": "ckeditor/ckeditor5-heading",
    "@ckeditor/ckeditor5-image": "ckeditor/ckeditor5-image",
    "@ckeditor/ckeditor5-link": "ckeditor/ckeditor5-link#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-list": "ckeditor/ckeditor5-list",
    "@ckeditor/ckeditor5-markdown-gfm": "ckeditor/ckeditor5-markdown-gfm",
    "@ckeditor/ckeditor5-paragraph": "ckeditor/ckeditor5-paragraph",
    "@ckeditor/ckeditor5-presets": "ckeditor/ckeditor5-presets",
    "@ckeditor/ckeditor5-theme-lark": "ckeditor/ckeditor5-theme-lark",
    "@ckeditor/ckeditor5-typing": "ckeditor/ckeditor5-typing",
    "@ckeditor/ckeditor5-ui": "ckeditor/ckeditor5-ui#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-undo": "ckeditor/ckeditor5-undo",
    "@ckeditor/ckeditor5-upload": "ckeditor/ckeditor5-upload#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-utils": "ckeditor/ckeditor5-utils#t/ckeditor5-dev/209",
    "@ckeditor/ckeditor5-widget": "ckeditor/ckeditor5-widget#t/ckeditor5-dev/209"
  }
}
```

**Comment by Reinmar** [comment]
Incorrect changes:

* https://github.com/ckeditor/ckeditor5-engine/compare/t/ckeditor5-dev/209#diff-3d31fd980a7d3a81f59e702c8ba2cbd9R614

   This stub will return the same instance on an on. It should return a different instance if it's called 2+.

* https://github.com/ckeditor/ckeditor5-link/compare/t/ckeditor5-dev/209#diff-2b6f216ce6cf947f7a48b79f59866735R46 and in a dozen of other places in this file.
  
   The same issue as above.

* https://github.com/ckeditor/ckeditor5-ui/compare/t/ckeditor5-dev/209#diff-ab35e9ac3f7822e474382a94e59cbde5L31

   The same issue as above.

* https://github.com/ckeditor/ckeditor5-utils/compare/t/ckeditor5-dev/209

   The same issue as above.

**Comment by pomek** [comment]
Fixed.

**Comment by Reinmar** [comment]
All green and looks fine.

**Pull request merged**